subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java

```diff
@@ -107,7 +110,43 @@ public void execute(@Nonnull String testClassName) {
         private void processAllTestClasses() {
             LauncherDiscoveryRequest discoveryRequest = createLauncherDiscoveryRequest(testClasses);
             TestExecutionListener executionListener = new JUnitPlatformTestExecutionListener(resultProcessor, clock, idGenerator);
-            launcherSession.getLauncher().execute(discoveryRequest, executionListener);
+            Launcher launcher = launcherSession.getLauncher();
+            if (spec.isDryRun()) {
+                TestPlan testPlan = launcher.discover(discoveryRequest);
+                executeDryRun(testPlan, executionListener);
+            } else {
+                launcher.execute(discoveryRequest, executionListener);
+            }
+        }
+    }
+
+    private void executeDryRun(TestPlan testPlan, TestExecutionListener listener) {
+        listener.testPlanExecutionStarted(testPlan);
+
+        for (TestIdentifier root : testPlan.getRoots()) {
+            dryRun(root, testPlan, listener);
+        }
+
+        listener.testPlanExecutionFinished(testPlan);
+    }
+
+    private void dryRun(TestIdentifier testIdentifier, TestPlan testPlan, TestExecutionListener listener) {
+        boolean isTest = testIdentifier.isTest();
+
+        if (!isTest) {
+            // if executionSkipped called, then executionStarted shouldn't be called for
+            // same testIdentifier. See TestExecutionListener for details.
+            listener.executionStarted(testIdentifier);
+        }
+
+        for (TestIdentifier child : testPlan.getChildren(testIdentifier)) {
+            dryRun(child, testPlan, listener);
+        }
+
+        if (isTest) {
+            listener.executionSkipped(testIdentifier, "Gradle test execution dry run");
+        } else {
+            listener.executionFinished(testIdentifier, TestExecutionResult.successful());
```

Comment:
I think the following should be equivalent but simpler:

```suggestion
        if (testIdentifier.isTest()) {
            listener.executionSkipped(testIdentifier, "Gradle test execution dry run");
        } else {
			listener.executionStarted(testIdentifier);
	        for (TestIdentifier child : testPlan.getChildren(testIdentifier)) {
	            dryRun(child, testPlan, listener);
	        }
            listener.executionFinished(testIdentifier, TestExecutionResult.successful());
        }
```